Ingest Cloudfront log files with Sidekiq and parse their contents

lozette opened the pull request:
Ingest Cloudfront log files from s3 in the same way as W3C IIS log files are ingested. Use a Sidekiq worker to ingest them once a day, and the new Cloudfront log parser to parse their contents.

Commit:
Add "host" and "url" to parsed log lines

These were omitted in a previous commit but are needed for parsing the log files

Commit:
Ingest Cloudfront log files along with IIS log files

This commit includes a small refactor to make the IIS & Cloudfront log parser
calls the same function

Commit:
Add a Sidekiq worker to ingest Cloudfront logs

Using the new Cloudfront log parser, ingest Cloudfront logs via Sidekiq.

Schedule the Sidekiq run the same way as the W3C logs are run, an hour later.

rjw1 commented on config/initializers/sidekiq.rb:
```diff
@@ -31,6 +31,11 @@
+      "daily_ingest_of_cloudfront_logs" => {
+        "class" => "IngestCloudfrontLogWorker",
+        "cron" => "0 20 * * *",
+        "args" => ENV["LOG_BUCKET_NAME"],
```
this env var should probably be named differently to the one for clf logs
Also i can see us having more than one bucket in the future.

lozette replied:
👍  Where are these set? If I rename them will it mess anything up?

rjw1 replied:
From our point of view nowhere currently. So changing them is fine.

Commit:
Allow the log ingest workers to handle multiple buckets

If the logs are stored in multiple buckets, the bucket names can be passed
to the Rake tasks as a comma-separated list, e.g

W3C_LOG_BUCKET_NAME=bucket-1
CLOUDFRONT_LOG_BUCKET_NAME=bucket-2,bucket-3,bucket-4

Either a single bucket or multiple buckets can be handled.

lozette commented:
I've amended the bucket arguments to handle a string of multiple buckets, e.g. `"bucket_1,bucket_2,bucket_3"` or just `"bucket"`

rjw1 reviewed (approved):
Looks good to me but a developer should probably look as well :)

Floppy commented on lib/transition/import/hits.rb:
```diff
@@ -128,7 +129,23 @@ def self.from_csv!(filename)
-        parsed_log_lines = parse_iis_w3c_log_file(filename: filename)
+        parsed_log_lines = parse_log_file(filename: filename, log_type: :iis)
+
+        # Create a temporary CSV file from the parsed CLF log lines
```
Seems like this code, that's duplicated in the following function, could be extracted into a different function to DRY it up, but I'm not worried about it enough to hold up merge.

Floppy commented on app/workers/ingest_cloudfront_log_worker.rb:
```diff
@@ -0,0 +1,48 @@
+require "services"
```
This seems extremely similar to the IIS log worker, and I think they could be DRYed up. All the specific stuff is in the `Transition::Import::Hits.from_cloudfront!` method, I think? But, as with my other comment, while it might be something to tidy up in future, it doesn't need to hold up the merge. Maybe open a ticket to DRY these up at some point in the future, maybe if we ever need to add another one.

lozette replied:
I have a few refactors I want to do on the IIS & Cloudfront parsers as we're extracting far more info from the log lines than we actually need.

rjw1 replied:
There is likely to be a couple more log formats (Default apache and nginx log formats. One of those would have helped UKRI bu they worked around that).

Floppy reviewed (approved):
A few things that could DRY up but they're not showstoppers.

![ok to go](https://user-images.githubusercontent.com/3565/102994597-43439900-4517-11eb-8e48-bbc7b3905db3.gif)
